server/s3_put_multiobject_copy_action.cc

```diff
@@ -146,16 +146,123 @@ void S3PutMultipartCopyAction::validate_multipart_partcopy_request() {
-      additional_object_metadata->get_content_length()) {
-    s3_copy_part_action_state = S3PutObjectActionState::validationFailed;
-    set_s3_error("InvalidRequest");
-    send_response_to_s3_client();
+  std::string range_header_value =
+      request->get_header_value("x-amz-copy-source-range");
+  source_object_size = additional_object_metadata->get_content_length();
+
+  if (range_header_value.empty()) {
+    // Range is not specified, read complete object
+    s3_log(S3_LOG_DEBUG, request_id, "Range is not specified\n");
+    if (MaxPartCopySourcePartSize < source_object_size) {
+      s3_copy_part_action_state = S3PutObjectActionState::validationFailed;
+      set_s3_error("InvalidRequest");
+      send_response_to_s3_client();
+    } else {
+      last_byte_offset_to_copy = source_object_size - 1;
+    }
   } else {
-    total_data_to_copy = additional_object_metadata->get_content_length();
-    next();
+    // parse the Range header value
+    // eg: bytes=0-1024 value
+    s3_log(S3_LOG_DEBUG, request_id, "Range found(%s)\n",
+           range_header_value.c_str());
+    if (!validate_range_header(range_header_value)) {
+      set_s3_error("InvalidRange");
+      send_response_to_s3_client();
+    }
+  }
+  total_data_to_copy = last_byte_offset_to_copy - first_byte_offset_to_copy;
+  s3_log(S3_LOG_DEBUG, request_id, "Validated range: %zu-%zu\n",
+         first_byte_offset_to_copy, last_byte_offset_to_copy);
+  s3_log(S3_LOG_DEBUG, request_id,
+         "Total data to copy in part [%d]: %zu bytes\n", part_number,
+         total_data_to_copy);
+  next();
+  s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
+}
+
+bool S3PutMultipartCopyAction::validate_range_header(
+    const std::string& range_value) {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+  // The header can consist of 'blank' character(s) only
+  if (std::find_if_not(range_value.begin(), range_value.end(), &::isspace) ==
+      range_value.end()) {
+    s3_log(S3_LOG_DEBUG, request_id,
+           "Range header consists of blank symbol(s) only");
+    return true;
+  }
+  // parse the Range header value
+  // eg: bytes=0-1024 value
+  std::size_t pos = range_value.find('=');
+  // return false when '=' not found
+  if (pos == std::string::npos) {
+    s3_log(S3_LOG_INFO, stripped_request_id, "Invalid range(%s)\n",
+           range_value.c_str());
+    return false;
+  }
+
+  std::string bytes_unit = S3CommonUtilities::trim(range_value.substr(0, pos));
+  std::string byte_range_set = range_value.substr(pos + 1);
+
+  // check bytes_unit has bytes string or not
+  if (bytes_unit != "bytes") {
+    s3_log(S3_LOG_INFO, stripped_request_id, "Invalid range(%s)\n",
+           range_value.c_str());
+    return false;
+  }
+
+  if (byte_range_set.empty()) {
+    // found range as bytes=
+    s3_log(S3_LOG_INFO, stripped_request_id, "Invalid range(%s)\n",
+           range_value.c_str());
+    return false;
+  }
+  // byte_range_set has multi range
```

Comment:
After extracting 'start' & 'end' bytes, we can just check that it has all numerals , like on line 247. so it will cover this extra ',' check too

Reply:
Some services (S3 or non S3 like Azure Blob) support multi range queries and we can specify multiple range in single requests separated by `,` like `bytes=100-200,400-500`
This check is done return specific error stating that we don't support multi range.

Reply:
Oh, got it. Looks fine then